scipy/stats/_stats_py.py

```diff
@@ -11008,3 +11008,126 @@ def first_order(t):
+
+
+def _prk(r, k):
+    return (-1)**(r-k)*special.binom(r, k)*special.binom(r+k, k)
+
+
+@_axis_nan_policy_factory(  # noqa: E302
+    _moment_result_object, n_samples=1, result_to_tuple=lambda x: (x,),
+    n_outputs=lambda kwds: kwds.get('n_moments', 4)
+)
+def lmoment(sample, n_moments=4, *, axis=0, sorted=False, standardize=True):
+    r"""Compute sample L-moments.
+
+    The L-moments of a probability distribution are summary statistics with
+    uses similar to those of conventional moments, but they are defined in
+    terms of the expected value of order statistics.
+    Sample L-moments are defined analogously to population L-moments, and
+    they can serve as estimators of population L-moments. They tend to be less
+    sensitive to extreme observations than conventional moments.
+
+    Parameters
+    ----------
+    sample : array_like
+        Array containing numbers whose expectile is desired.
+    n_moments : int
+        The number of moments to compute.
+    axis : int
+        The axis along which to compute L-moments.
+    sorted : bool, default=False
+        Whether `sample` is sorted in increasing order along `axis`.
+    standardize : bool, default=True
+        Whether to return L-moment ratios for orders 3 and higher.
+        L-moment ratios are analogous to standardized conventional
+        moments: they are the non-standardized L-moments divided
+        by the L-moment of order 2.
+
+    Returns
+    -------
+    lmoments : ndarray
+        The sample L-moments of orders 1 through `n_moments`.
+
+    See Also
+    --------
+    moment
+
+    References
+    ----------
+    D. Bilkova. "L-Moments and TL-Moments as an Alternative Tool of Statistical Data Analysis".
+    Journal of Applied Mathematics and Physics. 2014. :doi:`10.4236/jamp.2014.210104`
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from scipy import stats
+    >>> rng = np.random.default_rng(328458568356392)
+    >>> sample = rng.exponential(size=100000)
+    >>> stats.lmoment(sample)
+    array([1.00124272, 0.50111437, 0.3340092 , 0.16755338])
+
+    Note that the first four standardized population L-moments of the standard
+    exponential distribution are 1, 1/2, 1/3, and 1/6; the sample L-moments
+    provide reasonable estimates.
+
+    """
+    # Needs input validation and tests
+
+    args = lmoments_iv(sample, n_moments, axis, sorted, standardize)
+    sample, n_moments, axis, sorted, standardize = args
+
+    k = np.arange(n_moments, dtype=np.float64)
+    prk = _prk(np.expand_dims(k, list(range(1, sample.ndim+1))), k)
+    bk = _br(sample, r=k)
+
+    n = sample.shape[-1]
+    bk[..., n:] = 0  # remove NaNs due to n_moments > n
+
+    lmoms = np.sum(prk * bk, axis=-1)
+    if standardize and n_moments > 2:
+        lmoms[2:] /= lmoms[1]
+
+    lmoms[n:] = np.nan  # add NaNs where appropriate
+    if n_moments == 1:
+        lmoms = lmoms[0]
```

Comment:
It is tough to say whether this should be done. The point is moot if we replace the argument `n_moments` with something like `order`, which is an array-like of integer orders; the shape of the output will be determined by the shape of this array-like.